src/hooks/usePermit2.ts

```diff
@@ -0,0 +1,84 @@
+import { ContractTransaction } from '@ethersproject/contracts'
+import { PERMIT2_ADDRESS } from '@uniswap/permit2-sdk'
+import { CurrencyAmount, Token } from '@uniswap/sdk-core'
+import { useWeb3React } from '@web3-react/core'
+import useInterval from 'lib/hooks/useInterval'
+import ms from 'ms.macro'
+import { useCallback, useEffect, useMemo, useState } from 'react'
+import { ApproveTransactionInfo } from 'state/transactions/types'
+
+import { PermitSignature, usePermitAllowance, useUpdatePermitAllowance } from './usePermitAllowance'
+import { useTokenAllowance, useUpdateTokenAllowance } from './useTokenAllowance'
+
+export enum PermitState {
+  UNKNOWN,
+  PERMIT_NEEDED,
+  PERMITTED,
+}
+
+export interface Permit {
+  state: PermitState
+  signature?: PermitSignature
+  callback?: () => Promise<{
+    response: ContractTransaction
+    info: ApproveTransactionInfo
+  } | void>
+}
+
+export default function usePermit(amount?: CurrencyAmount<Token>, spender?: string): Permit {
+  const { account } = useWeb3React()
+  const tokenAllowance = useTokenAllowance(amount?.currency, account, PERMIT2_ADDRESS)
+  const updateTokenAllowance = useUpdateTokenAllowance(amount, PERMIT2_ADDRESS)
+
+  const allowanceData = usePermitAllowance(amount?.currency, spender)
+  const [permitAllowance, setPermitAllowance] = useState(allowanceData?.amount)
```

Comment:
is this state redundant? It looks like it's just being kept in sync with `allowanceData.amount`

Reply:
It is, but it's cleared should the expiration pass. It is used to prevent re-running the memoization at the frequency of the interval timer.